models/demos/whisper/tt/ttnn_optimized_functional_whisper.py

```diff
@@ -0,0 +1,672 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+import ttnn
+import torch
+import transformers
+from typing import Optional
+from torch.nn import functional as F
+from ttnn.model_preprocessing import preprocess_linear_weight, preprocess_linear_bias
+
+
+WHISPER_DTYPE = ttnn.bfloat8_b
+
+
+def dropout(hidden_states, p, training):
+    # ignored for inference
+    return hidden_states
+
+
+# The split_query_key_value_and_split_heads requires the query to have the same volume as the key and values
+# This is not the case however for whisper so we currently cannot swap out calculate_key_values below
+# def calculate_key_values(config, query_states, key_value_states, *, parameters):
+#     fused_kv = key_value_states @ parameters.key_value.weight + parameters.key_value.bias
+#     head_size = config.d_model // config.encoder_attention_heads
+#     batch_size, *_, _, two_times_hidden_size = fused_kv.shape.with_tile_padding()
+#     hidden_size = two_times_hidden_size // 2
+#     encoder_attention_heads = hidden_size // head_size
+#     query_states, key_states, value_states = ttnn.transformer.split_query_key_value_and_split_heads(
+#         query_states,
+#         kv_input_tensor=fused_kv,
+#         num_heads=encoder_attention_heads,
+#         memory_config=WHISPER_MEMORY_CONFIG,
+#     )
+#     key_states = ttnn.permute(key_states, (0, 1, 3, 2))
+#     return query_states, key_states, value_states
+
+
+def calculate_key_values(config, key_value_states, *, parameters, whisper_memory_config):
+    bsz, tgt_len, hidden_size = key_value_states.shape
+    bsz, tgt_len_padded, _ = key_value_states.shape.with_tile_padding()
+    head_size = hidden_size // config.encoder_attention_heads
+
+    fused_qkv = ttnn.linear(
+        key_value_states,
+        parameters.weight,
+        bias=parameters.bias,
+        memory_config=whisper_memory_config,
+    )
+    dtype = fused_qkv.dtype
+    device = fused_qkv.device()
+
+    # fused_qkv = ttnn.to_layout(fused_qkv, layout=ttnn.ROW_MAJOR_LAYOUT)
+    # fused_qkv = ttnn.from_device(fused_qkv)
+    # fused_qkv = ttnn.reshape(fused_qkv, (bsz, tgt_len, config.encoder_attention_heads, 2, head_size))
+    # # Without Split: 0.84 pcc
+    # key_states = ttnn.reshape(fused_qkv, (bsz, tgt_len, config.encoder_attention_heads, head_size * 2))[..., :head_size]
+    # value_states = ttnn.reshape(fused_qkv, (bsz, tgt_len, config.encoder_attention_heads, head_size * 2))[..., head_size:]
+
+    # key_states = ttnn.to_device(key_states, device)
+    # key_states = ttnn.to_layout(key_states, ttnn.TILE_LAYOUT)
+    # key_states = ttnn.permute(key_states, (0, 2, 3, 1))
+
+    # value_states = ttnn.to_device(value_states, device)
+    # value_states = ttnn.to_layout(value_states, ttnn.TILE_LAYOUT)
+    # value_states = ttnn.permute(value_states, (0, 2, 1, 3))
```

Comment:
can these comments be removed?

Reply:
I've updated the file removing comments.